server/src/main/java/org/opensearch/index/translog/transfer/TranslogCheckpointTransferSnapshot.java

```diff
@@ -164,7 +164,6 @@ public TranslogCheckpointTransferSnapshot build() throws IOException {
             translogTransferSnapshot.setMinTranslogGeneration(highestGenMinTranslogGeneration);
 
             assert this.primaryTerm == highestGenPrimaryTerm : "inconsistent primary term";
-            assert this.generation == highestGeneration : " inconsistent generation ";
```

Comment:
why do we need to remove this ?

Reply:
As part of the combined upload operation, we acquire the write lock, release it and then acquire read lock, and then upload. In this process it is possible that the generation can get rolled over after releasing the write lock and before acquiring the read lock. But most importantly, we have acquired the translog generation log and translog / ckp deletion would not happen.

Reply:
Are you suggesting we will end up uploading a higher generation translog?

Reply:
Can write lock -> unlock -> read lock transition lead to underlying referenced files to closed?
Would it be much cleaner if we performed a `TranslogCheckpointTransferSnapshot #build` and set the deletion policy to acquire lock on min generation within the write lock itself for better guarantees?

Reply:
> Are you suggesting we will end up uploading a higher generation translog?

It's possible that we can upload a higher generation of the reader (tlog/ckp). 

> Can write lock -> unlock -> read lock transition lead to underlying referenced files to closed?

No, that would not happen for 2 reasons - 1. trim unreferenced readers ensure relies on uploaded generation check 2. we acquire the translogGenLock here -> https://github.com/opensearch-project/OpenSearch/blob/b5cc00222816429ed673d9fb46c434cae1be862b/server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java#L350
But I do see that we are more protected by the cleanup of unreferenced readers logic. 

> Would it be much cleaner if we performed a TranslogCheckpointTransferSnapshot #build and set the deletion policy to acquire lock on min generation within the write lock itself for better guarantees?

This makes sense totally. Let me test this out and push the change.

Reply:
The translog gen lock acquisition only happens at upload which is when we move to read lock. The concern I have is there are discrepancies in the prepare and the upload part, which this assertion has been able to catch which we are disabling here which is concerning since this establishes that there could be races between when you downgrade the writer lock to a read lock non-atomically. While there are ways to do this atomically I see potential conflicts due to races in how we switch between locks

While there

Reply:
https://github.com/opensearch-project/OpenSearch/issues/10117 - we will be fixing this.